packages/core/src/shared/utilities/functionUtils.ts

```diff
@@ -77,6 +77,20 @@ export function memoize<T, U extends any[]>(fn: (...args: U) => T): (...args: U)
     return (...args) => (cache[args.map(String).join(':')] ??= fn(...args))
 }
 
+/**
+ * Generalization of the {@link memoize} method that accepts async methods, and allows user to pass mapping from keys to args.
```

Comment:
can we extend the existing memoize()? also probably want direct test coverage of this

Reply:
I was trying to maintain its type signature. Making this one return an async method would cause a lot of unrelated changes. However, no longer relevant since I refactored away from this.

Reply:
> Making this one return an async method

would an "override" work?